app/vlstorage/lastnoptimization.go

```diff
@@ -103,15 +103,15 @@ func getLastNQueryResults(qctx *logstorage.QueryContext, limit uint64) ([]logRow
 			return rowsFound, nil
 		}
 
-		// The number of found rows is below the limit. This means the [start ... end] time range
+		// The number of found rows is below the limit. This means the [start ... end) time range
 		// doesn't cover the needed logs, so it must be extended.
 		// Append the found rows to rowsFound, adjust n, so it doesn't take into account already found rows
 		// and adjust the time range to search logs at [start-(end/2-start/2) ... start).
 		rowsFound = append(rowsFound, rows...)
 		n -= uint64(len(rows))
 
 		d := end/2 - start/2
-		end = start - 1
+		end = start
```

Comment:
@vadimalekseev , could provide a link to the commit in the v1.36.0 release, which introduced the bug, with a clear explanation why this commit created this bug.

Probably, that commit introduced bugs in other places of VictoriaLogs codebase too.

Reply:
When VictoriaLogs applies sorting, it requests `2*limit` logs from each vlstorage. If more than `2*limit` logs are returned, we narrow the time range to get fewer than `2*limit` logs.

If vlstorage returns N < limit logs, vlselect keeps these N and tries to fetch the remaining limit-N logs.

The problem occurs when vlstorage returns fewer than limit logs: at this point, we subtract 1 nanosecond from the end of the query, and `/select/logsql/query` also subtracts 1 nanosecond. As a result, 2 nanoseconds are lost instead of one.

Because of this, it is difficult to predict which logs will be excluded from the query - it depends on the time range and the data returned by vlstorage.